plugins/historydata/ua_history_data_backend_memory.c

```diff
@@ -603,3 +606,268 @@ UA_HistoryDataBackend_Memory_clear(UA_HistoryDataBackend *backend)
+
+    size_t storeEnd =
+        backend->getEnd(server, backend->context, sessionId, sessionContext, nodeId);
+    size_t startIndex;
+    size_t endIndex;
+    UA_Boolean addFirst;
+    UA_Boolean addLast;
+    UA_Boolean reverse;
+    size_t _resultSize = getResultSize_service_Circular(
+        backend, server, sessionId, sessionContext, nodeId, start, end,
+        numValuesPerNode == 0 ? 0 : numValuesPerNode + (UA_UInt32)skip, returnBounds,
+        &startIndex, &endIndex, &addFirst, &addLast, &reverse);
+
+    *resultSize = _resultSize - skip;
+    if(*resultSize > maxSize) {
+        *resultSize = maxSize;
+    }
+    UA_DataValue *outResult =
+        (UA_DataValue *)UA_Array_new(*resultSize, &UA_TYPES[UA_TYPES_DATAVALUE]);
+    if(!outResult) {
+        *resultSize = 0;
+        return UA_STATUSCODE_BADOUTOFMEMORY;
+    }
+    *result = outResult;
+    size_t counter = 0;
+    if(addFirst) {
+        if(skip == 0) {
+            outResult[counter].hasStatus = true;
+            outResult[counter].status = UA_STATUSCODE_BADBOUNDNOTFOUND;
+            outResult[counter].hasSourceTimestamp = true;
+            if(start == LLONG_MIN) {
+                outResult[counter].sourceTimestamp = end;
+            } else {
+                outResult[counter].sourceTimestamp = start;
+            }
+            ++counter;
+        }
+    }
+    UA_ByteString backendOutContinuationPoint;
+    UA_ByteString_init(&backendOutContinuationPoint);
+    if(endIndex != storeEnd && startIndex != storeEnd) {
+        size_t retval = 0;
+
+        size_t valueSize = *resultSize - counter;
+        if(valueSize + skip > _resultSize - addFirst - addLast) {
+            if(skip == 0) {
+                valueSize = _resultSize - addFirst - addLast;
+            } else {
+                valueSize = _resultSize - skip - addLast;
+            }
+        }
+
+        UA_StatusCode ret = UA_STATUSCODE_GOOD;
+        if(valueSize > 0)
+
+            ret = backend->copyDataValues(
+                server, backend->context, sessionId, sessionContext, nodeId, startIndex,
+                endIndex, reverse, valueSize, range, releaseContinuationPoints,
+                &backendContinuationPoint, &backendOutContinuationPoint, &retval,
+                &outResult[counter]);
+        if(ret != UA_STATUSCODE_GOOD) {
+
+            UA_Array_delete(outResult, *resultSize, &UA_TYPES[UA_TYPES_DATAVALUE]);
+            *result = NULL;
+            *resultSize = 0;
+            return ret;
+        }
+        counter += retval;
+    }
+    if(addLast && counter < *resultSize) {
+
+        outResult[counter].hasStatus = true;
+        outResult[counter].status = UA_STATUSCODE_BADBOUNDNOTFOUND;
+        outResult[counter].hasSourceTimestamp = true;
+        if(start == LLONG_MIN &&
+           storeEnd != backend->firstIndex(server, backend->context, sessionId,
+                                           sessionContext, nodeId)) {
+            outResult[counter].sourceTimestamp =
```

Comment:
Please document that code.
Why do you subtract one second at the end?

Reply:
@jpfr  This part of the code is identical to the "normal" implementation. I did not change it because is not the core part of the circular buffer historizing algorithm.
Maybe @peter-rustler-basyskom know tha answer.

Meanwhile I implemented an example and a minimal unit test for the circular buffer, as well as following your suggestions on cosmetics and code documentation.